### PR TITLE
Deep Extension Resolution

### LINKED ISSUE #50: Getting a 'should have been resolved BEFORE serialization' error
I'm getting an error that prevents me from using reusable components with leaf.

The code is the following:
```
#extend("Templates/base"):
    #export("body"):
        #for(challenge in challenges):
            #extend("Components/single-challenge-total"):#endextend
        #endfor
    #endexport
#endextend
```

It results in the following error:
```
{"error":true,"reason":"syntax extend(\"Components\/single-challenge-total\") should have been resolved BEFORE serialization"}
```

I've tried this with leaf-kit 1.0.0-rc.1.6 and leaf 4.0.0-rc.1.2

### PR DESCRIPTION
**Pull request by tdotclare**
### DEEP RESOLVE ###

* Significantly improved LeafAST resolution
* Syntax items now recursively rebuild themselves to allow deeply-nested #extend tags to properly resolve
* Simplify LeafAST resolution steps, correct tail recursion
* Adds appropriate test cases

This is an ugly but nonetheless functional update to the referenced template resolution; ASTs will now recursively examine ALL Syntax objects during loading and resolution to inline templates no matter how deeply buried.

I used tons of inouts on associated value Enums, don't @ me. This is not ideal in terms of code beauty but it does what it purports to.

Resolves #50

### CODE REVIEW

File: Sources/LeafKit/LeafAST.swift [hunk 1]
```diff
@@ -33,75 +31,53 @@ public struct LeafAST: Hashable {
...
+                    ast[pos] = .raw(buffer)
+                    ast.remove(at: ast.index(after: pos) )
+                } else { pos = ast.index(after: pos) }
             } else { pos = ast.index(after: pos) }
         }
         
-        // update refs as new externals may have been introduced through extension
-        updateRefs()
+        flat = unresolvedRefs.isEmpty ? true : false
+        if firstRun, flat { rawAST = nil }
```
Comment (gwynne) [inline Sources/LeafKit/LeafAST.swift:80]: For a conditional like this, prefer `&&` to the comma syntax, which is really intended more for use with optional bindings

File: Sources/LeafKit/LeafSyntax.swift [hunk 2]
```diff
@@ -291,3 +291,112 @@ extension Syntax: CustomStringConvertible {
...
+
+extension Syntax {
+    mutating func inlineRefs(_ externals: [String: LeafAST], _ new: inout Syntax?) -> Set<String> {
+        var changed = false
+        var rf = Set<String>()
+
+        switch self {
+            case .conditional(let co):
+                var tail: Syntax.Conditional? = co
+                while tail != nil {
```
Comment (gwynne) [inline Sources/LeafKit/LeafSyntax.swift:303]: Change the `var tail` to name it `currentTail`, then use `while let tail = currentTail {` as the loop condition, then you can get rid of all the `!` uses in the loop body and change the end to `currentTail = tail.next`.

File: Sources/LeafKit/LeafAST.swift (hunk 1 above)
Comment (tdotclare) [reply Sources/LeafKit/LeafAST.swift]: Done

File: Sources/LeafKit/LeafSyntax.swift (hunk 2 above)
Comment (tdotclare) [reply Sources/LeafKit/LeafSyntax.swift]: Done

**Comment by tanner0101** [comment]
These changes are now available in [1.0.0-rc.1.8](https://github.com/vapor/leaf-kit/releases/tag/1.0.0-rc.1.8)

**Pull request merged**